cassiopeia/tools/tree.py

```diff
@@ -0,0 +1,227 @@
+                edge_length = self.get_edge_length(v, child)
+                if child in leaves:
+                    subtree_newick = subtree_newick_representation(child)
+                else:
+                    subtree_newick = (
+                        "(" + subtree_newick_representation(child) + ")"
+                    )
+                    if print_internal_nodes:
+                        subtree_newick += format_node(child)
+                # Add edge length
+                subtree_newick = subtree_newick + ":" + str(edge_length)
+                if print_pct_of_mutated_characters_along_edge:
+                    # Also add number of mutations
+                    number_of_unmutated_characters_in_parent = self.get_state(
+                        v
+                    ).count("0")
+                    number_of_mutations_along_edge = self.get_state(v).count(
+                        "0"
+                    ) - self.get_state(child).count("0")
+                    pct_of_mutated_characters_along_edge = (
+                        number_of_mutations_along_edge
+                        / (number_of_unmutated_characters_in_parent + 1e-100)
+                    )
+                    subtree_newick = (
+                        subtree_newick + "[&&NHX:muts="
+                        f"{self._fmt(pct_of_mutated_characters_along_edge)}]"
+                    )
+                subtrees_newick.append(subtree_newick)
+            newick = ",".join(subtrees_newick)
+            return newick
+
+        root = self.root()
+        res = "(" + subtree_newick_representation(root) + ")"
+        if print_internal_nodes:
+            res += format_node(root)
+        res += ");"
+        return res
+
+    def _fmt(self, x: float):
+        return "%.2f" % x
+
+    def reconstruct_ancestral_states(self):
+        r"""
+        Reconstructs ancestral states with maximum parsimony.
+        """
+        root = self.root()
+
+        def dfs(v: int) -> None:
+            children = self.children(v)
+            n_children = len(children)
+            if n_children == 0:
+                return
+            for child in children:
+                dfs(child)
+            children_states = [self.get_state(child) for child in children]
+            n_characters = len(children_states[0])
+            state = ""
+            for character_id in range(n_characters):
+                states_for_this_character = set(
+                    [
+                        children_states[i][character_id]
+                        for i in range(n_children)
+                    ]
+                )
+                if len(states_for_this_character) == 1:
+                    state += states_for_this_character.pop()
+                else:
+                    state += "0"
+            self.set_state(v, state)
+            if v == root:
+                # Reset state to all zeros!
+                self.set_state(v, "0" * n_characters)
+
+        dfs(root)
```

Comment:
Richard wrote a function that does this in `cassiopeia.solver.solver_utilities`. You might just want to take that because it takes into account missing data, etc.

Reply:
Yes, you are right! I just found it easier to code this up myself then to figure out how to get the right inputs to `annotate_ancestral_characters`, because e.g. the Tree class I am using holds character vectors as a `str`, not as a `List[int]` :( Probably a bad design choice of mine? This is one thing I wanted to discuss! I left the character vector as a `str` because it met all my needs, and changing it to `List[int]` meant changing the code prematurely. (We definitely want to have only one implementation of parsimony reconstruction when we merge, the question is what's the API/where it lives.)

Reply:
Makes sense! This is definitely something out of scope of your branch length estimator.

But, yes I feel that using lists to represent the character states of a sample is more desirable. One major reason for this is the separation of characters in a string is not always intuitive (e.g., if you have double-digit or triple-digit state representations). One can get around this by setting a delimiter (e.g., the `|` symbol) but if you always end up splitting the string to change something and then rejoining, why not just ditch the string data structure altogether for this?